jujugui/static/gui/src/app/components/terminal/terminal.js

```diff
@@ -83,70 +79,77 @@ class Terminal extends React.Component {
       // Terminado sends a "disconnect" message when the process it's running
       // exits. When we receive that, we close the terminal.
       if (resp['0'] === 'disconnect') {
-        this.setOpened(false);
+        this.close();
       }
       if (resp.code === 'ok' && resp.message === 'session is ready') {
-        const term = new XTerm();
         term.terminadoAttach(ws);
-        term.open(ReactDOM.findDOMNode(this)
-          .querySelector('.juju-shell__terminal-container'));
-        this.term = term;
+        term.writeln('connected to temporary workspace.\n');
       }
     };
-    this.ws = ws;
   }
 
-  /**
-    Destroy the terminal window and close the WebSocket connection to the
-    Juju shell service.
-  */
-  stopTerm() {
+  componentWillUnmount() {
     this.term.destroy();
     this.term = null;
     this.ws.close();
     this.ws = null;
   }
 
+  setSize(size) {
+    this.setState({size: size}, () => {
+      if (this.term) {
+        this.term.fit();
+      }
+    });
+  }
+
+  close() {
+    this.props.changeState({
+      terminal: null
+    });
+  }
+
   render() {
-    if (this.state.opened) {
-      return (
-        <Lightbox close={this.setOpened.bind(this, false)}>
-          <div className="juju-shell__terminal-container"></div>
-        </Lightbox>
-      );
+    const state = this.state;
+    const terminalClassNames = classnames(
+      'juju-shell__terminal', {
+        'juju-shell__terminal--min': state.size === 'min'
+      });
+    const styles = {};
+    if (state.size === 'max') {
+      styles.height = window.innerHeight - 250 + 'px';
     }
-    const props = this.props;
-    const address = props.address;
-    const classes = classNames(
-      'model-actions__import',
-      'model-actions__button',
-      {'model-actions__button-disabled': !address}
-    );
     return (
-      <span className={classes}
-        onClick={address && this.setOpened.bind(this, true)}
-        role="button"
-        tabIndex="0">
-        <SvgIcon name="code-snippet_24"
-          className="model-actions__icon"
-          size="16" />
-        <span className="tooltip__tooltip--below">
-          <span className="tooltip__inner tooltip__inner--up">
-            Juju shell
-          </span>
-        </span>
-      </span>
+      <div className="juju-shell">
+        <div className="juju-shell__header">
+          <span className="juju-shell__header-label">Juju Shell</span>
+          <div className="juju-shell__header-actions">
+            <span onClick={this.setSize.bind(this, 'min')}>
+              <SvgIcon name="minimize-bar_16" size="16" />
+            </span>
+            <span onClick={this.setSize.bind(this, 'max')}>
+              <SvgIcon name="maximize-bar_16" size="16" />
+            </span>
+            <span onClick={this.close.bind(this)}>
```

Comment:
These clickable spans should probably have `tabIndex="0" role="button"` so they can be accessed by the keyboard.

Reply:
good call